keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1435 @@
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+We introduce a new Job API field, called `.spec.backoffLimitPerIndex`.
+When set it limits the number of retries, counted independently for all indexes.
+
+Additionally, we propose the `.spec.maxFailedIndexes` to control
+the maximal number of failed indexes. Once the number is exceeded the entire
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
+
+const (
+  // This is an action which might be taken on a pod failure - mark the
+  // Job's index as failed to avoid restarts within this index. This action
+  // can only be used when backoffLimitPerIndex is set.
+  PodFailurePolicyActionFailIndex PodFailurePolicyAction = "FailIndex"
+  ...
+)
+...
+
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+  ...
+  // Specifies the limit for the number of retries within an
+  // index before marking this index as failed. When enabled the number of
+  // failures per index is kept in the pod's
+  // batch.kubernetes.io/job-index-failure-count annotation. It can only
+  // be set when Job's completionMode=Indexed, and the Pod's restart
+  // policy is Never.
+  // When specified, then the backoffLimit field is defaulted to the max int32
+  // value. Also, once specified it cannot be set to nil.
+  // +optional
+  BackoffLimitPerIndex *int32
+
+  // Specifies the maximal number of failed indexes before marking the Job as
+  // failed, when backoffLimitPerIndex is set. Once the number of failed
+  // indexes exceeds this number the entire Job is marked as Failed and its
+  // execution is terminated. When left as nil the job continues execution of
+  // all of its indexes and is marked with the `Complete` Job condition.
+  // It can only be specified when backoffLimitPerIndex is set.
+  // The value is might be up to 10^5 when completions is <= 10^5, or 10^4 when
+  // completions > 10^5.
+  // +optional
+  MaxFailedIndexes *int32
+  ...
+}
+
+type JobStatus struct {
+  ...
+
+  // FailedIndexes holds the failed indexes when backoffLimitPerIndex is set.
+  // The indexes are represented in the text format analogous as for the
+  // `completedIndexes` field, ie. they are kept as decimal integers
+  // separated by commas. The numbers are listed in increasing order. Three or
+  // more consecutive numbers are compressed and represented by the first and
+  // last element of the series, separated by a hyphen.
+  // For example, if the failed indexes are 1, 3, 4, 5 and 7, they are
+  // represented as "1,3-5,7".
+  // +optional
+  FailedIndexes *string
```

Comment:
I know that we technically don't increase the order of of magnitude in the size of the object, but for large objects it's important to ensure that the load from those will be somewhat limited.

I was pointing into that in the past for IndexedJobs that we need to ensure some batching around object (in particular status) updates and I think it happened, but can you double check and ensure that it is/will be the case?

Reply:
Job controller maintains a queue of tasks (`syncJob` invocations). So events such as pod updates don't directly trigger syncJob, but put the owning job into the queue. The queue is processed by workers (default: 5) which are triggered every 1s. So the batching is by deduplication of items in the queue within the 1s intervals.